base_layer/core/src/mempool/utx_pool/utx_pool.rs

```diff
@@ -0,0 +1,316 @@
+        assert!(utx_pool.contains_key(&utx1.body.kernels[0].excess_sig).unwrap());
+        assert!(!utx_pool.contains_key(&utx2.body.kernels[0].excess_sig).unwrap());
+        assert!(utx_pool.contains_key(&utx3.body.kernels[0].excess_sig).unwrap());
+        assert!(utx_pool.contains_key(&utx4.body.kernels[0].excess_sig).unwrap());
+        assert!(utx_pool.contains_key(&utx5.body.kernels[0].excess_sig).unwrap());
+        // Retrieve the set of highest priority unspent transactions
+        let selected_utxs = utx_pool.highest_priority_utxs(3).unwrap();
+        assert_eq!(selected_utxs.len(), 3);
+        assert_eq!(selected_utxs[0].body.kernels[0].fee, MicroTari(1000));
+        assert_eq!(selected_utxs[1].body.kernels[0].fee, MicroTari(500));
+        assert_eq!(selected_utxs[2].body.kernels[0].fee, MicroTari(500));
+
+        assert!(utx_pool.check_status().unwrap());
+    }
+
+    #[test]
+    fn test_remove_published_txs() {
+        let utx1 = create_test_tx(MicroTari(500));
+        let utx2 = create_test_tx(MicroTari(100));
+        let utx3 = create_test_tx(MicroTari(1000));
+        let utx4 = create_test_tx(MicroTari(200));
+        let utx5 = create_test_tx(MicroTari(500));
+        let utx6 = create_test_tx(MicroTari(750));
+
+        let mut utx_pool = UTxPool::new(UTxPoolConfig { storage_capacity: 10 });
+        utx_pool
+            .insert(vec![
+                utx1.clone(),
+                utx2.clone(),
+                utx3.clone(),
+                utx4.clone(),
+                utx5.clone(),
+            ])
+            .unwrap();
+        // utx6 should not be added to utx_pool as it is an unknown transactions that was included in the block by
+        // another node
+
+        let published_block = create_test_block(vec![utx1.clone(), utx3.clone(), utx5.clone()]);
+        let _ = utx_pool.remove_published(&published_block);
+
+        assert!(!utx_pool.contains_key(&utx1.body.kernels[0].excess_sig).unwrap());
+        assert!(utx_pool.contains_key(&utx2.body.kernels[0].excess_sig).unwrap());
+        assert!(!utx_pool.contains_key(&utx3.body.kernels[0].excess_sig).unwrap());
+        assert!(utx_pool.contains_key(&utx4.body.kernels[0].excess_sig).unwrap());
+        assert!(!utx_pool.contains_key(&utx5.body.kernels[0].excess_sig).unwrap());
+        assert!(!utx_pool.contains_key(&utx6.body.kernels[0].excess_sig).unwrap());
+
+        assert!(utx_pool.check_status().unwrap());
+    }
+
+    #[test]
+    fn test_remove_orphaned_txs() {
+        let utx1 = create_test_tx(MicroTari(500));
+        let utx2 = create_test_tx(MicroTari(100));
+        let utx3 = create_test_tx(MicroTari(1000));
+        let utx4 = create_test_tx(MicroTari(200));
+        let mut utx5 = create_test_tx(MicroTari(500));
+        let mut utx6 = create_test_tx(MicroTari(750));
+        // utx1 and utx5 have a shared input. Also, utx3 and utx6 have a shared input
```

Comment:
These types of TX are called double-spends usually, and aren't technically orphans. Hence my confusion earlier. In fact they shouldn't go to the orphan pool, since in the absence of a re-org, they can never be spent.

Reply:
Behaviour was changed to rather discard double spends instead of handling them as orphaned transactions.

Reply:
💯